File: fantasticauth-sdk-ios/Sources/Vault/Vault.swift
import Foundation
import Combine

/// The main Vault SDK class.
///
/// Use this class to configure and interact with the Vault platform.
/// All interactions are thread-safe and use Swift concurrency.
///
/// Example usage:
/// ```swift
/// import Vault
///
/// // Configure the SDK
/// Vault.configure(
///     apiUrl: "https://api.vault.dev",
///     tenantId: "my-tenant"
/// )
///
/// // Sign in
/// let auth = VaultAuth()
/// let session = try await auth.signIn(email: "user@example.com", password: "password")
/// ```
public final class Vault {
    
    // MARK: - Singleton
    
    /// The shared Vault instance.
    public static let shared = Vault()
    
    // MARK: - Configuration
    
    /// The current configuration.
    public private(set) var configuration: Configuration?
    
    /// Configuration for the Vault SDK.
    public struct Configuration {
        /// The API URL for Vault.
        public let apiUrl: String
        
        /// The tenant identifier.
        public let tenantId: String
        
        /// Optional API key for server-side operations.
        public let apiKey: String?
        
        /// Whether to enable debug logging.
        public let debugMode: Bool
        
        /// Request timeout in seconds.
        public let timeout: TimeInterval
        
        /// The current SDK version.
        public let sdkVersion: String = "1.0.0"
        
        /// Creates a new configuration.
        ///
        /// - Parameters:
        ///   - apiUrl: The API URL for Vault (e.g., "https://api.vault.dev")
        ///   - tenantId: Your tenant identifier
        ///   - apiKey: Optional API key for admin operations
        ///   - debugMode: Enable debug logging (default: false)
        ///   - timeout: Request timeout in seconds (default: 30)
        public init(
            apiUrl: String,
            tenantId: String,
            apiKey: String? = nil,
            debugMode: Bool = false,
            timeout: TimeInterval = 30
        ) {
            self.apiUrl = apiUrl.trimmingCharacters(in: .init(charactersIn: "/"))
            self.tenantId = tenantId
            self.apiKey = apiKey
            self.debugMode = debugMode
            self.timeout = timeout
        }
    }
    
    // MARK: - Services
    
    /// The authentication service.
    public lazy var auth: VaultAuth = {
        VaultAuth(apiClient: apiClient, tokenStore: tokenStore)
    }()
    
    /// The session manager.
    public lazy var session: VaultSession = {
        VaultSession(tokenStore: tokenStore, apiClient: apiClient)
    }()
    
    /// The OAuth service.
    public lazy var oauth: VaultOAuth = {
        VaultOAuth(apiClient: apiClient, tokenStore: tokenStore)
    }()
    
    /// The biometric authentication service.
    public lazy var biometric: VaultBiometric = {
        VaultBiometric(tokenStore: tokenStore)
    }()
    
    /// The organizations service.
    public lazy var organizations: VaultOrganizations = {
        VaultOrganizations(apiClient: apiClient, session: session)
    }()
    
    // MARK: - Internal Services
    
    /// The API client.
    internal lazy var apiClient: APIClient = {
        guard let config = configuration else {
            fatalError("Vault must be configured before use. Call Vault.configure() first.")
        }
        return APIClient(configuration: config)
    }()
    
    /// The token store.
    internal let tokenStore: TokenStore = TokenStore()
    
    /// The logger.
    internal let logger = VaultLogger()
    
    // MARK: - Combine Publishers
    
    /// Publisher that emits when the session state changes.
    public var sessionPublisher: AnyPublisher<SessionState, Never> {
        session.statePublisher
    }
    
    // MARK: - Initialization
    
    private init() {}
    
    // MARK: - Configuration
    
    /// Configures the Vault SDK.
    ///
    /// This method must be called before using any other Vault functionality.
    /// It's safe to call this method multiple times - the last configuration will be used.
    ///
    /// - Parameter configuration: The configuration to use.
    public static func configure(_ configuration: Configuration) {
        shared.configuration = configuration
        shared.logger.debugMode = configuration.debugMode
        shared.logger.log("Vault SDK configured with tenant: \(configuration.tenantId)")
        
        // Pre-warm the API client
        _ = shared.apiClient
    }
    
    /// Configures the Vault SDK with convenience parameters.
    ///
    /// - Parameters:
    ///   - apiUrl: The API URL for Vault
    ///   - tenantId: Your tenant identifier
    ///   - apiKey: Optional API key for admin operations
    ///   - debugMode: Enable debug logging (default: false)
    public static func configure(
        apiUrl: String,
        tenantId: String,
        apiKey: String? = nil,
        debugMode: Bool = false
    ) {
        configure(Configuration(
            apiUrl: apiUrl,
            tenantId: tenantId,
            apiKey: apiKey,
            debugMode: debugMode
        ))
    }
    
    // MARK: - Utility
    
    /// Returns the current SDK version.
    public static var version: String {
        "1.0.0"
    }
    
    /// Checks if the SDK has been configured.
    public var isConfigured: Bool {
        configuration != nil
    }
    
    /// Resets the SDK state (useful for testing).
    public func reset() {
        Task {
            await session.signOut()
        }
        logger.log("Vault SDK reset")
    }
}

// MARK: - Session State

/// Represents the current session state.
public enum SessionState: Equatable {
    /// No user is signed in.
    case unauthenticated
    
    /// A user is signed in.
    case authenticated(VaultUser)
    
    /// The session is being refreshed.
    case refreshing
}

File: fantasticauth-sdk-android/src/main/java/dev.fantasticauth/sdk/Vault.kt
package dev.vault.sdk

import android.content.Context
import dev.vault.sdk.auth.VaultAuth
import dev.vault.sdk.biometric.VaultBiometric
import dev.vault.sdk.network.APIClient
import dev.vault.sdk.organizations.VaultOrganizations
import dev.vault.sdk.session.VaultSession
import dev.vault.sdk.utils.VaultLogger

/**
 * Main entry point for the Vault Android SDK.
 * 
 * Example usage:
 * ```kotlin
 * // In your Application.onCreate() or MainActivity
 * Vault.configure(
 *     apiUrl = "https://api.vault.dev",
 *     tenantId = "my-tenant"
 * )
 * 
 * // Then use throughout your app
 * val auth = VaultAuth()
 * val session = VaultSession()
 * ```
 */
object Vault {
    
    private var _config: VaultConfig? = null
    private var _context: Context? = null
    private var _isInitialized = false
    
    /**
     * SDK version string
     */
    const val VERSION = BuildConfig.SDK_VERSION
    
    /**
     * Current SDK configuration
     * @throws IllegalStateException if SDK is not initialized
     */
    val config: VaultConfig
        get() = _config ?: throw IllegalStateException(
            "Vault SDK not initialized. Call Vault.configure() first."
        )
    
    /**
     * Application context
     * @throws IllegalStateException if SDK is not initialized
     */
    val context: Context
        get() = _context ?: throw IllegalStateException(
            "Vault SDK not initialized. Call Vault.configure() first."
        )
    
    /**
     * Whether the SDK has been initialized
     */
    val isInitialized: Boolean
        get() = _isInitialized
    
    /**
     * Configure the Vault SDK with required parameters.
     * Should be called once, typically in Application.onCreate() or MainActivity.onCreate()
     * 
     * @param context Application context
     * @param apiUrl Base URL for the Vault API
     * @param tenantId Your tenant identifier
     * @param timeout Request timeout in seconds (default: 30)
     * @param enableLogging Whether to enable debug logging (default: false in release)
     */
    @JvmStatic
    @JvmOverloads
    fun configure(
        context: Context,
        apiUrl: String,
        tenantId: String,
        timeout: Long = 30,
        enableLogging: Boolean = BuildConfig.DEBUG
    ) {
        if (_isInitialized) {
            VaultLogger.w("Vault SDK already initialized. Ignoring duplicate configure() call.")
            return
        }
        
        _context = context.applicationContext
        _config = VaultConfig(
            apiUrl = apiUrl.removeSuffix("/"),
            tenantId = tenantId,
            timeout = timeout
        )
        
        VaultLogger.configure(enableLogging)
        APIClient.initialize(_config!!)
        
        _isInitialized = true
        VaultLogger.i("Vault SDK initialized - v$VERSION")
    }
    
    /**
     * Reset the SDK configuration. Useful for testing or switching tenants.
     */
    @JvmStatic
    fun reset() {
        _config = null
        _context = null
        _isInitialized = false
        APIClient.reset()
        VaultLogger.i("Vault SDK reset")
    }
    
    /**
     * Create a new Auth instance
     */
    @JvmStatic
    fun auth(): VaultAuth = VaultAuth()
    
    /**
     * Create a new Session instance
     */
    @JvmStatic
    fun session(): VaultSession = VaultSession()
    
    /**
     * Create a new Biometric instance
     */
    @JvmStatic
    fun biometric(context: Context): VaultBiometric = VaultBiometric(context)
    
    /**
     * Create a new Organizations instance
     */
    @JvmStatic
    fun organizations(): VaultOrganizations = VaultOrganizations()
}

/**
 * Configuration for the Vault SDK
 */
data class VaultConfig(
    val apiUrl: String,
    val tenantId: String,
    val timeout: Long
) {
    /**
     * Full API base URL including tenant
     */
    val baseUrl: String
        get() = "$apiUrl/v1/tenants/$tenantId"
}
